lib/weakhash/weakhash.go

```diff
@@ -8,111 +8,121 @@ package weakhash
-	if ir == nil || len(hashesToFind) == 0 {
-		return nil, nil
-	}
+	r  io.ReaderAt
+	br *bufio.Reader
 
-	r := bufio.NewReader(ir)
-	hf := adler32.New()
+	// Last matching hash value.
+	match uint32
+}
 
-	n, err := io.CopyN(hf, r, int64(size))
-	if err == io.EOF {
-		return nil, nil
-	}
-	if err != nil {
-		return nil, err
-	}
-	if n != int64(size) {
-		return nil, io.ErrShortBuffer
-	}
+type Reader interface {
+	io.ReaderAt
+	io.Reader
+}
 
-	offsets := make(map[uint32][]int64)
-	for _, hashToFind := range hashesToFind {
-		offsets[hashToFind] = make([]int64, 0, maxWeakhashFinderHits)
+// NewFinder returns a Finder that initially has no hashes.
+// Call Add before Next to add hashes.
+//
+// The buffer buf will be filled for each match found. Its length is taken
+// to be the block size.
+func NewFinder(r Reader, buf []byte) *Finder {
+	f := &Finder{
+		buf:    buf,
+		hashes: make(map[uint32]struct{}),
+		hf:     adler32.New(),
+		r:      r,
+		br:     bufio.NewReader(r),
 	}
 
-	var i int64
-	var hash uint32
-	for {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		default:
-		}
-
-		hash = hf.Sum32()
-		if existing, ok := offsets[hash]; ok && len(existing) < maxWeakhashFinderHits {
-			offsets[hash] = append(existing, i)
-		}
-		i++
-
-		bt, err := r.ReadByte()
-		if err == io.EOF {
-			break
-		} else if err != nil {
-			return offsets, err
-		}
-		hf.Roll(bt)
-	}
-	return offsets, nil
+	return f
 }
 
-func NewFinder(ctx context.Context, ir io.ReadSeeker, size int, hashesToFind []uint32) (*Finder, error) {
-	offsets, err := Find(ctx, ir, hashesToFind, size)
-	if err != nil {
-		return nil, err
+// Add adds the hash h to f.
+func (f *Finder) Add(h uint32) { f.hashes[h] = struct{}{} }
+
+// Err returns the last error encountered by Next, if any.
+// EOF is not considered an error.
+func (f *Finder) Err() error {
+	switch f.err {
+	case io.EOF, io.ErrUnexpectedEOF:
+		return nil
+	default:
+		return f.err
 	}
-
-	return &Finder{
-		reader:  ir,
-		size:    size,
-		offsets: offsets,
-	}, nil
 }
 
-type Finder struct {
-	reader  io.ReadSeeker
-	size    int
-	offsets map[uint32][]int64
+// Match returns the hash and offset of the last match found by Next.
+func (f *Finder) Match() (h uint32, offset int64) {
+	return f.match, f.offset - int64(len(f.buf))
 }
 
-// Iterate iterates all available blocks that matches the provided hash, reads
-// them into buf, and calls the iterator function. The iterator function should
-// return whether it wishes to continue iterating.
-func (h *Finder) Iterate(hash uint32, buf []byte, iterFunc func(int64) bool) (bool, error) {
-	if h == nil || hash == 0 || len(buf) != h.size {
-		return false, nil
+// Next returns true if it can find a match for any of f's hashes,
+// false if it encounters either an error or EOF.
+//
+// When Next has returned true, the contents of the block found are in the
+// buffer handed to NewFinder.
+func (f *Finder) Next() bool {
```

Comment:
I'd return an error only (with EOF meaning all good, but finished), or `(bool, error)`. Then you can drop the `Error` method and it's immediately evident on the loop whether an error occurred, without the api user needing to think of calling `Error` afterwards.

Reply:
The API design is a conscious decision: it makes the loop check a single expression. Note that the error from finder.Err is special, in that it is logged and then ignored. Regarding EOF, that isn't an error, it's an expected condition; there are no partial results to handle as in io.Reader.Read. bufio.Scanner discards it for the same reason.